web-client/public/js/upload.js

```diff
@@ -83,6 +91,121 @@ export const upload = function () {
+        if (simTimepoints) {
+            optimizationParameters.data["simulation_timepoints"] = simTimepoints;
+        }
+        optimizationParameters.data.species = "Saccharomyces cerevisiae";
+        optimizationParameters.data["taxon_id"] = 559292;
+        return optimizationParameters;
+    };
+
+    const handleExpressionDataAndExport = (route, extension, sheetType, source, finalExportSheets) => {
+        if (source === "userInput" && grnState.workbook.expression) {
+            // make sure that the optimization parameters sheet is actually properly formatted
+            if (finalExportSheets["optimization_parameters"]) {
+                finalExportSheets["optimization_parameters"] = updateOptimizationParameters(finalExportSheets);
+            }
+            grnState.workbook.exportSheets = finalExportSheets;
+            exportExcel(route, extension, sheetType);
+        } else {
+            // expression source is from database so lets query her up
+            for (let sheet in finalExportSheets.expression) {
+                startLoadingIcon();
+                queryExpressionDatabase({
+                    type:"ExpressionData",
+                    dataset: `${source}_${sheet.replace("_log2_expression", "")}`,
+                    genes : grnState.workbook.genes.map(x => {
+                        return x.name;
+                    }).join(",")
+                }).then(function (response) {
+                    finalExportSheets.expression[sheet] = response;
+                    if (finalExportSheets.expression[sheet]) {
+                        stopLoadingIcon();
+                        if ( !Object.values(finalExportSheets.expression).includes(null) ) {
+                            // we have all of the expression sheets so lets initilize the export process
+                            Object.keys(finalExportSheets.expression).forEach((sheet) => {
+                                // make sure that the sheets we queried are populated with the correct data
+                                if (!(finalExportSheets.expression[sheet].data && finalExportSheets.expression[sheet].timePoints)) {
+                                    // if the resulting query doesn't contains both the timePoint data and
+                                    // the gene data then don't export it. If not don't :)
+                                    finalExportSheets.expression[sheet] = null;
+                                }
+                            });
+                            if (finalExportSheets["optimization_parameters"] === null) {
+                                finalExportSheets["optimization_parameters"] = updateOptimizationParameters(finalExportSheets);
+                            }
+                            grnState.workbook.exportSheets = finalExportSheets;
+                            exportExcel(route, extension, sheetType);
+                        }
+                    }
+                }).catch(function (error) {
```

Comment:
To keep the indentation manageable, this `function (response)` (and why isn’t it just `response =>`?) can be defined externally thus allowing this line to become something like `.then(expressionDataHandler)` or whatever name it gets

Reply:
Found some issues after running `npm run lint`. Revised by fixing indentation and changing method call to be: `.then((expressionData) => expressionDataHandler(expressionData, sheet, route, extension, sheetType, finalExportSheets))` so that helper method gets all necessary information.